### PR TITLE
 feat(clouddriver): wait for instances down when server group is disabled

### PR DESCRIPTION
**Pull request by dreynaud**
This is meant to prevent the situation where some instances in a disabled server group can still have a health status Up, resulting in scaleDownCluster skipping that server group.

### CODE REVIEW

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy [hunk 1]
```diff
@@ -62,39 +62,39 @@ class WaitForClusterDisableTask extends AbstractWaitForClusterWideClouddriverTas
...
+    }?.type
+
+    return platformHealthType ? platformHealthType : healthProviderNamesByPlatform[getCloudProvider(stage)]
+  }
+
   @Override
   boolean isServerGroupOperationInProgress(Stage stage,
                                            List<Map> interestingHealthProviderNames,
                                            Optional<TargetServerGroup> serverGroup) {
+    // TODO(dreynaud): this can probably be changed to interestingHealthProviderNames == null || isEmpty()
```
Comment (jrsquared) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: So..... change it?

Comment (dreynaud) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: *probably* :)

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/AbstractWaitForClusterWideClouddriverTask.groovy [hunk 2]
```diff
@@ -53,7 +53,7 @@ abstract class AbstractWaitForClusterWideClouddriverTask extends AbstractCloudPr
   protected TaskResult emptyClusterResult(Stage stage,
                                           AbstractClusterWideClouddriverTask.ClusterSelection clusterSelection,
                                           Map cluster) {
-    throw new IllegalStateException("No ServerGroups found in cluster $clusterSelection")
+    throw new IllegalStateException("no server groups found in cluster $clusterSelection")
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/AbstractWaitForClusterWideClouddriverTask.groovy]: It was more consistent with the other `IllegalStateException` that can be thrown, `'no cluster details found'` something something

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy [hunk 3]
```diff
@@ -62,39 +62,39 @@ class WaitForClusterDisableTask extends AbstractWaitForClusterWideClouddriverTas
     return taskResult
   }
 
+  private def getPlatformHealthType(Stage stage, TargetServerGroup targetServerGroup) {
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: This is just moving existing code in a named method, I really don't know what the return value is. `targetServerGroup.instances` is a `List<Map>`

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+import spock.lang.Specification
+import spock.lang.Subject
+import spock.lang.Unroll
+
+import static com.netflix.spinnaker.orca.ExecutionStatus.RUNNING
+import static com.netflix.spinnaker.orca.ExecutionStatus.SUCCEEDED
+
+import java.util.concurrent.TimeUnit
+
+class WaitForClusterDisableTaskSpec extends Specification {
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: I see, I had the EditorConfig plugin disabled

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+    def oortHelper = Mock(OortHelper)
+    def katoService = Mock(KatoService)
+    def objectMapper = new ObjectMapper()
+
+    @Shared def region = "theregion"
+    @Shared def clusterName = "thecluster"
+
+
+    @Shared
+    ServerGroupCreator aCreator = Stub(ServerGroupCreator) {
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: Not sure about the Mock vs Stub distinction, I stole this from a neighboring test

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        getCloudProvider() >> "aCloud"
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: What value do you suggest? Using a random value has the advantage of being explicitly something that does not matter in the system being tested

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/AbstractWaitForClusterWideClouddriverTask.groovy (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/AbstractWaitForClusterWideClouddriverTask.groovy]: 👍

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        def stage = new Stage(Execution.newPipeline("orca"), "test", [
+                cluster                                     : clusterName,
+                credentials                                 : "theaccount",
+                "deploy.server.groups"                      : [
+                        (dsgregion): ["$clusterName-$oldServerGroup".toString()]
+                ],
+                (desiredPct ? "desiredPercentage" : "blerp"): desiredPct,
+                interestingHealthProviderNames              : interestingHealthProviderNames
+        ])
+        stage.setStartTime(System.currentTimeMillis() - TimeUnit.HOURS.toMillis(1))
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: ha, yes! There is a minimum wait time (90s I think). Maybe I can parameterize it to exercise the different behaviors too

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: It seems reasonable to find out what the actual values and do a cast if nothing else

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy (hunk 4 above)
[the remark this answers is not in the document]
Comment (ajordens) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: 💥

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+        task.waitForRequiredInstancesDownTask = new WaitForRequiredInstancesDownTask()
+
+        when:
+        TaskResult result = task.execute(stage)
+
+        then:
+        result.getStatus() == status
+
+        where:
+        dsgregion | oldSGDisabled | desired | desiredPct | interestingHealthProviderNames | platformHealthState || status
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: I started with longer names, but I squeezed some of them as the data table is getting out of hand (in terms of line length)

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy (hunk 6 above)
[the remark this answers is not in the document]
Comment (ajordens) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: Typically I (and others from Netflix) would use`aws` ... I can see the value of using a completely random value _but_ we haven't been doing that normally.

Alternatively I could see `"cloudProvider"` also being an appropriate return value for `getCloudProvider()` in a test case. At least if something fails that's perhaps every so slightly more indicative of where to look that `"aCloud"`?

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+                health: [[healthClass: 'platform', type: 'platformHealthType', state: platformHealthState]],
+                healthState: null,
+                zone: 'thezone']
+    }
+
+    def serverGroup(name, region, Map other) {
+        return [
+                name  : name,
+                region: region,
+                health: null
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: indeed

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy (hunk 7 above)
[the remark this answers is not in the document]
Comment (ajordens) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: Ahh isn't that the hokey pokey.

MINIMUM_WAIT_TIME_MS while being all nice and capitalized isn't actually a constant and couple be explicitly set.

That class is all sorts of weird with a `@Autowired` constructor as well as some `@Autowired` and `@Value` class variables.

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+        ] + other
+    }
+
+    @Unroll
+    def "status=#status when dsgregion=#dsgregion, oldServerGroupDisabled=#oldServerGroupDisabled, desiredPercentage=#desiredPercentage, interestingHealthProviderNames=#interestingHealthProviderNames"() {
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: good point about `oldServerGroupDisabled`, same pb with `desiredPercentage` (again a result of me squeezing variable names a bit after the fact). I agree that the unrolled names become a bit unreadable, in fact I stopped adding them after a while (e.g. no `platformHealthState`). I'll condense them a little bit and maybe keep the most meaningful ones (`disabled` and `desiredPercentage` in particular as they have a big impact on what is being tested)

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+
+        when:
+        TaskResult result = task.execute(stage)
+
+        then:
+        IllegalStateException e = thrown()
+        e.message.startsWith(message)
+
+        where:
+        clusterData << [
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: I anticipated bigger sizes for the values, in particular for `cluster`, but they turned out to be manageable

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy (hunk 6 above)
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: `"cloudProvider"` sounds good to me 👍

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+    @Shared
+    ServerGroupCreator aCreator = Stub(ServerGroupCreator) {
+        getCloudProvider() >> "aCloud"
+        isKatoResultExpected() >> false
+        getOperations(_) >> [["aOp": "foo"]]
+    }
+
+    @Subject def task = new WaitForClusterDisableTask([aCreator])
+
+    def instance(name, platformHealthState = 'Unknown') {
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: so... `Map`?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: how would I go about finding out the type if the application logic doesn't provide it? Isn't it a bit backwards to have tests more strongly typed than the app logic?

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy [hunk 13]
```diff
@@ -62,39 +62,39 @@ class WaitForClusterDisableTask extends AbstractWaitForClusterWideClouddriverTas
...
-      }
-
-      return !(platformHealthType && interestingHealthProviderNames == [platformHealthType])
     }
+
+    // Even if the server group is disabled, we want to make sure instances are down
+    // to prevent downstream stages (e.g. scaleDownCluster) from having to deal with disabled-but-instances-up server groups
+    def targetServerGroup = serverGroup.get()
+    if (targetServerGroup.isDisabled() || stage.context.desiredPercentage) {
+      return !waitForRequiredInstancesDownTask.hasSucceeded(stage, targetServerGroup as Map, targetServerGroup.getInstances(), interestingHealthProviderNames)
```
Comment (dreynaud) [inline orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: *this* here is the crucial business logic change. We used to return early on `targetServerGroup.isDisabled()`, but now I make it use the same logic as when we have a `desiredPercentage`, piggybacking on `waitForRequiredInstancesDownTask.hasSucceeded`. Does this seem correct @ajordens @cfieber

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+
+import static com.netflix.spinnaker.orca.ExecutionStatus.RUNNING
+import static com.netflix.spinnaker.orca.ExecutionStatus.SUCCEEDED
+
+import java.util.concurrent.TimeUnit
+
+class WaitForClusterDisableTaskSpec extends Specification {
+    def oortHelper = Mock(OortHelper)
+    def katoService = Mock(KatoService)
+    def objectMapper = new ObjectMapper()
```
Comment (dreynaud) [inline orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: going to clean some of these up, turns out we don't need an `objectMapper` or a `katoService`

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: this is the real code tho ... I'd just set a break point and take a quick look

File: orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+                name: clusterName,
+                serverGroups: [
+                        serverGroup("$clusterName-v050".toString(), "us-west-1", [:]),
+                        serverGroup("$clusterName-v051".toString(), "us-west-1", [:]),
+                        serverGroup("$clusterName-$newServerGroup".toString(), region, [:]),
+                        serverGroup("$clusterName-$oldServerGroup".toString(), region, [
+                                disabled: oldSGDisabled,
+                                capacity: [desired: desired],
+                                instances: [
+                                        instance('i-1', platformHealthState),
```
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: I was hoping I could get away without testing the extra health providers logic in the interest of time 😭

[the remark this answers is not in the document]
Comment (ajordens) [reply orca-clouddriver/src/test/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTaskSpec.groovy]: :shrug: 

That's fair ... but in that case I'd just make `waitForRequiredInstancesDownTask` a `Mock` and all you need to do is ensure that it's called with the right parameters.

Assuming there's coverage elsewhere we don't need to waste a lot of lines here doubling up.

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy (hunk 13 above)
Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: Lars did have a comment around whether the non-desiredPercentage scenario (which is what you're actually doing here).

Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: It doesn't appear if previously we were doing something specific to handle a case that the server group is not disabled yet?

Comment (dreynaud) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: Old logic was:
- asg disabled -> we're done
- asg not disabled, but we have desiredPercentage -> wait for required instances down
- asg not disabled, no desiredPercentage -> done if we are only supposed to consider platform health (regardless of platform health status, which I find odd)

New logic is:
- asg disabled or we have desiredPercentage -> wait for required instances down
- asg not disabled, no desiredPercentage -> unchanged

CC @lwander

Comment (ajordens) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: 👍

File: orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy (hunk 3 above)
[the remark this answers is not in the document]
Comment (dreynaud) [reply orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/cluster/WaitForClusterDisableTask.groovy]: `health:[[type:Amazon, healthClass:platform, state:Unknown]`, it seems `type` in this case was just the `Amazon` string, but I guess there's no guarantee that it's always going to be the case

**Comment by dthomson25** [comment]
This PR seems related to the open issue: https://github.com/spinnaker/spinnaker/issues/1506.  @dreynaud 's description fits that issue's case where the scale down task skips the old server groups.

**Comment by dreynaud** [comment]
@dthomson25 yes indeed, I think this should address spinnaker/spinnaker#1506

**Comment by shraykay** [comment]
just experienced this on our spinnaker implementation, any chance we can get this merged?

**Comment by ajordens** [comment]
I’ve not run this PR locally so let’s give it a good whirl in prestaging.

**Comment by dreynaud** [comment]
@shraykay trying to land this today 🤞

**Comment by shraykay** [comment]
awesome - thanks guys!

**Comment by shraykay** [comment]
i tried running this nightly along with the latest clouddriver, however my cluster groups without load balancers are still not scaling down. are there any other services that require updating?

**Comment by dreynaud** [comment]
No, this should be self-contained. If your issue is always reproducible, it
may have a different cause

2018-04-15 23:32 GMT-07:00 Shray Kumar <notifications@github.com>:

**Pull request merged**